# Sentimental Analysis using Bernoulli and Multinomial Navie Bayes on ecommerce dataset 

In [60]:
import pandas as pd
df=pd.read_csv("ecommerceDataset.csv")
df.head

<bound method NDFrame.head of         Category                                               Text
0      Household  SAF 'Floral' Framed Painting (Wood, 30 inch x ...
1      Household  SAF 'UV Textured Modern Art Print Framed' Pain...
2      Household  SAF Flower Print Framed Painting (Synthetic, 1...
3      Household  Incredible Gifts India Wooden Happy Birthday U...
4      Household  Pitaara Box Romantic Venice Canvas Painting 6m...
...          ...                                                ...
44453        NaN                                                NaN
44454        NaN                                                NaN
44455        NaN                                                NaN
44456        NaN                                                NaN
44457        NaN                                                NaN

[44458 rows x 2 columns]>

In [61]:
df.shape

(44458, 2)

In [62]:
df=df.dropna()
df
df.shape

(499, 2)

In [63]:
y=df.iloc[:,:-1]
x=df.iloc[:,-1]
x

0      SAF 'Floral' Framed Painting (Wood, 30 inch x ...
1      SAF 'UV Textured Modern Art Print Framed' Pain...
2      SAF Flower Print Framed Painting (Synthetic, 1...
3      Incredible Gifts India Wooden Happy Birthday U...
4      Pitaara Box Romantic Venice Canvas Painting 6m...
                             ...                        
494    Decdeal DIY Handmade Needlework Counted Cross ...
495    Anchor Fly with Me Stitch Kit A range of over ...
496    Decdeal Anself DIY Handmade Needlework Counted...
497    Borosil Carafe with Strainer in Lid, 1.5 Litre...
498    Global Home Microfiber Soft Korean Floral Embo...
Name: Text, Length: 499, dtype: object

In [64]:
y

Category
0      Household
1      Household
2      Household
3      Household
4      Household
..           ...
494  Electronics
495  Electronics
496  Electronics
497  Electronics
498  Electronics

[499 rows x 1 columns]

In [65]:
#Custom Stop_words
custom=["the","is","or","this","can","that",",",".","could","do","does","x"]

In [66]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.naive_bayes import BernoulliNB,MultinomialNB,GaussianNB
vector1=CountVectorizer(binary=True,stop_words="english")
vector2=CountVectorizer(binary=False,stop_words="english")
vector3=CountVectorizer(binary=True,stop_words=custom)
vector4=CountVectorizer(binary=False,stop_words=custom)

In [67]:
x1=vector1.fit_transform(x)
x2=vector2.fit_transform(x)
x3=vector3.fit_transform(x)
x4=vector4.fit_transform(x)

In [68]:
from sklearn.model_selection import train_test_split
xtrain1,xtest1,ytrain1,ytest1=train_test_split(x1,y,random_state=1,test_size=0.3)
xtrain2,xtest2,ytrain2,ytest2=train_test_split(x2,y,random_state=1,test_size=0.3)
xtrain3,xtest3,ytrain3,ytest3=train_test_split(x3,y,random_state=1,test_size=0.3)
xtrain4,xtest4,ytrain4,ytest4=train_test_split(x4,y,random_state=1,test_size=0.3)

In [69]:
nnmodel=BernoulliNB()
mnmodel=MultinomialNB()
nnmodel1=BernoulliNB()
mnmodel1=MultinomialNB()

In [70]:
nnmodel.fit(xtrain1,ytrain1)
nnmodel1.fit(xtrain3,ytrain3)

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\utils\validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\utils\validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)


BernoulliNB()

In [71]:
mnmodel.fit(xtrain2,ytrain2)
mnmodel1.fit(xtrain4,ytrain4)

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\utils\validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\utils\validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)


MultinomialNB()

In [72]:
pred1=nnmodel.predict(xtest1)
pred2=mnmodel.predict(xtest2)
pred3=nnmodel1.predict(xtest3)
pred4=mnmodel1.predict(xtest4)

In [73]:
from sklearn.metrics import accuracy_score
print("Accuracy using default stop words..BernoulliNB:\n",accuracy_score(ytest1,pred1))
print("Accuracy using custom stop words..BernoulliNB:\n",accuracy_score(ytest3,pred3))

Accuracy using default stop words..BernoulliNB:
 0.76
Accuracy using custom stop words..BernoulliNB:
 0.74


In [74]:
print("Accuracy using default stop words..MultinomialNB:\n",accuracy_score(ytest2,pred2))
print("Accuracy using custom stop words..MultinomialNB:\n",accuracy_score(ytest4,pred4))

Accuracy using default stop words..MultinomialNB:
 0.8533333333333334
Accuracy using custom stop words..MultinomialNB:
 0.88


In [75]:
from sklearn.pipeline import make_pipeline
from sklearn.feature_extraction.text import TfidfVectorizer
xtrain,xtest,ytrain,ytest=train_test_split(x,y,random_state=21,test_size=0.3)
model=make_pipeline(TfidfVectorizer(stop_words="english"),MultinomialNB())
model.fit(xtrain,ytrain)

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\utils\validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)


Pipeline(steps=[('tfidfvectorizer', TfidfVectorizer(stop_words='english')),
                ('multinomialnb', MultinomialNB())])

In [76]:
model1=make_pipeline(TfidfVectorizer(stop_words=custom),MultinomialNB())
model1.fit(xtrain,ytrain)

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\utils\validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)


Pipeline(steps=[('tfidfvectorizer',
                 TfidfVectorizer(stop_words=['the', 'is', 'or', 'this', 'can',
                                             'that', ',', '.', 'could', 'do',
                                             'does', 'x'])),
                ('multinomialnb', MultinomialNB())])

In [77]:
pred5=model.predict(xtest)
pred6=model1.predict(xtest)
print("Accuracy_Score using default stop words:\n",accuracy_score(ytest,pred5))
print("Accuracy_Score using custom stop words:\n",accuracy_score(ytest,pred6))

Accuracy_Score using default stop words:
 0.8733333333333333
Accuracy_Score using custom stop words:
 0.8533333333333334


# To conclude:The highest accuracy score 0.8733 using MultinomialNB with Tfidf vectorizor removing the default stop_words from the model as compare to other vectorizer technique. 

In [78]:
from sklearn.metrics import confusion_matrix
print("Confusion metrics using default stop words:\n",confusion_matrix(ytest,pred5))
print("Confusion metrics using custom stop words:\n",confusion_matrix(ytest,pred6))

Confusion metrics using default stop words:
 [[73  0  0  0]
 [10 28  0  1]
 [ 0  0  0  1]
 [ 4  3  0 30]]
Confusion metrics using custom stop words:
 [[73  0  0  0]
 [11 27  0  1]
 [ 0  0  0  1]
 [ 6  3  0 28]]


# Saving the model

In [79]:
import joblib
joblib.dump(model,"MultinomialNB_model_using TFIDF vectorizer")

['MultinomialNB_model_using TFIDF vectorizer']

In [80]:
loaded_model = joblib.load('MultinomialNB_model_using TFIDF vectorizer')

In [88]:
new_text=["Transform your home with quality products from Home Essentials Store. Happy shopping!"]
prediction=loaded_model.predict(new_text)
print("Output:\n",prediction)

Output:
 ['Books']
